apps/emqx_durable_storage/src/emqx_ds_beamformer.erl

```diff
@@ -295,590 +418,952 @@ poll(Node, ReturnAddr, Shard, Iterator, Opts = #{timeout := Timeout}) ->
+                ok ->
+                    Unmatched;
+                Error ->
+                    ?tp(
+                        error,
+                        emqx_ds_beamformer_delegate_error,
+                        #{shard => DBShard, worker => Worker, error => Error}
+                    ),
+                    Reqs ++ Unmatched
+            end
         end,
-        ReqsByNode
+        [],
+        ByWorker
     ),
-    NFulfilled > 0.
-
--spec pack(
-    fun((Iterator, emqx_ds:message_key()) -> Iterator),
-    emqx_ds:message_key(),
-    [{ItKey, Iterator}],
-    [{emqx_ds:message_key(), emqx_types:message()}]
-) -> beam(ItKey, Iterator).
-pack(UpdateIterator, NextKey, Reqs, Batch) ->
-    Pack = [{Key, mk_mask(Reqs, Elem), Msg} || Elem = {Key, Msg} <- Batch],
-    UpdatedIterators =
-        lists:map(
-            fun(#poll_req{it = It0, return_addr = RAddr}) ->
-                {ok, It} = UpdateIterator(It0, NextKey),
-                {RAddr, It}
-            end,
-            Reqs
-        ),
-    #beam{
-        iterators = UpdatedIterators,
-        pack = Pack
-    }.
-
-split([], _Pack, _N, Acc) ->
-    Acc;
-split([{ItKey, It} | Rest], Pack, N, Acc0) ->
-    Msgs = [
-        {MsgKey, Msg}
-     || {MsgKey, Mask, Msg} <- Pack,
-        is_member(N, Mask)
-    ],
-    case Msgs of
-        [] -> logger:warning("Empty batch ~p", [ItKey]);
-        _ -> ok
-    end,
-    Acc = [{ItKey, {ok, It, Msgs}} | Acc0],
-    split(Rest, Pack, N + 1, Acc).
-
--spec is_member(non_neg_integer(), dispatch_mask()) -> boolean().
-is_member(N, Mask) ->
-    <<_:N, Val:1, _/bitstring>> = Mask,
-    Val =:= 1.
-
--spec mk_mask([poll_req(_ItKey, _Iterator)], {emqx_ds:message_key(), emqx_types:message()}) ->
-    dispatch_mask().
-mk_mask(Reqs, Elem) ->
-    mk_mask(Reqs, Elem, <<>>).
+    {Unmatched =:= [], D#d{pending = Unmatched}}.
 
-mk_mask([], _Elem, Acc) ->
-    Acc;
-mk_mask([#poll_req{msg_matcher = Matcher} | Rest], {Key, Message} = Elem, Acc) ->
-    Val =
-        case Matcher(Key, Message) of
-            true -> 1;
-            false -> 0
-        end,
-    mk_mask(Rest, Elem, <<Acc/bitstring, Val:1>>).
-
-filter_candidates(Reqs, Messages) ->
-    lists:partition(
-        fun(#poll_req{msg_matcher = Matcher}) ->
-            lists:any(
-                fun({MsgKey, Msg}) -> Matcher(MsgKey, Msg) end,
-                Messages
-            )
-        end,
-        Reqs
-    ).
+remove_subscription(
+    SubId, D = #d{dbshard = DBShard, stuck_sub_tab = SubTab, monitor_tab = MTab, pending = Pending}
+) ->
+    ets:delete(fc_tab(DBShard), SubId),
+    ets:delete(SubTab, SubId),
+    case ets:take(owner_tab(DBShard), SubId) of
+        [#owner_tab{owner = Owner, mref = MRef}] ->
+            %% Remove client monitoring:
+            demonitor(MRef, [flush]),
+            ets:delete(MTab, MRef),
+            %% Remove stuck subscription (if stuck):
+            ets:delete(disowned_sub_tab(DBShard), SubId),
```

Comment:
Since we get `stuck_sub_tab = SubTab` from the gen_statem data, it seems that such table is already [what's in the gvar](https://github.com/emqx/emqx/blob/fe6e9966ec532166549afefd03d895fd5729cceb/apps/emqx_durable_storage/src/emqx_ds_beamformer.erl#L726-L728).  If really so, we have already deleted the subscription in the 2nd line of this function.